zilliqa/tests/it/main.rs

```diff
@@ -278,6 +277,110 @@ impl Network {
     }
 
+    pub fn restart(&mut self) {
+        // We copy the data dirs from the original network, and re-use the same private keys.
+
+        // Note: the tempdir object has to be held in the vector or the OS
+        // will delete it when it goes out of scope.
+        let mut options = CopyOptions::new();
+        options.copy_inside = true;
+
+        // Collect the keys from the validators
+        let keys = self.nodes.iter().map(|n| n.secret_key).collect::<Vec<_>>();
+
+        let validator = (
+            keys[0].node_public_key(),
+            keys[0].to_libp2p_keypair().public().to_peer_id(),
+        );
+        let genesis_committee = vec![validator];
+
+        let (nodes, mut receivers): (Vec<_>, Vec<_>) = keys
+            .into_iter()
+            .enumerate()
+            .map(|(i, key)| {
+                // Copy the persistence over
+                let new_data_dir = tempfile::tempdir().unwrap();
+
+                info!("Copying data dir over");
+
+                if let Ok(mut entry) = fs::read_dir(self.nodes[i].dir.as_ref().unwrap().path()) {
+                    let entry = entry.next().unwrap().unwrap();
+                    info!("Copying {:?} to {:?}", entry, new_data_dir);
+
+                    copy(entry.path(), new_data_dir.path(), &options).unwrap();
+                } else {
+                    warn!("Failed to copy data dir over");
+                }
+
+                node(
+                    genesis_committee.clone(),
+                    None,
+                    key,
+                    i,
+                    Some(new_data_dir),
+                    self.genesis_address,
+                )
+                .unwrap()
+            })
+            .unzip();
+
+        for node in &nodes {
+            trace!(
+                "Node {}: {} (dir: {})",
+                node.index,
+                node.peer_id,
+                node.dir.as_ref().unwrap().path().to_string_lossy(),
+            );
+        }
+
+        let (resend_message, receive_resend_message) =
+            mpsc::unbounded_channel::<(PeerId, Option<PeerId>, Message)>();
+        let receive_resend_message = UnboundedReceiverStream::new(receive_resend_message).boxed();
+        receivers.push(receive_resend_message);
+
+        self.nodes = nodes;
+        self.receivers = receivers;
+        self.resend_message = resend_message;
+
+        // Now trigger a timeout in all of the nodes until we see network activity again
+        // this could of course spin forever, but the test itself should time out.
+        loop {
+            for node in &self.nodes {
+                if node.inner.lock().unwrap().handle_timeout() {
+                    return;
+                }
+                zilliqa::time::advance(Duration::from_millis(1));
+            }
+        }
+    }
+
+    // Drop the first message in each node queue with N% probability per tick
+    pub async fn randomly_drop_messages_then_tick(&mut self, failure_rate: f64) {
```

Comment:
If block production still works with dropped messages, this makes me wonder whether we can just integrate this logic into `tick()` and run _all_ our tests on an unreliable network (rather than just having this special case method)?

Worth a try I think (possibly in a different PR though if you think it'll delay this one).

Reply:
I wouldn't favour doing that because it would make debugging the tests more annoying, plus it takes significantly longer for the network to stabilize.

On the other hand if done manually could potentially reveal stuff like forks created by flaky networking not being handled correctly I guess. 

I think on balance, your test should try and test only the functionality you want it to test, and not fail for some other kindof unrelated reason (for example you change networking and some other test now sometimes fails).

Reply:
That's fair. I would argue that most, if not all, tests should ultimately be testing the network under non-ideal conditions - it's why we have the randomisation in the first place, for example.
But you make a good point that it will make it much harder to debug failures.